### DISCUSSION TITLE
Page switch does not scroll to top for scrollable container (`Link`'s `scroll` not working) [`v13.1.7`]

### DISCUSSION THREAD

**Discussion by bennettdams** [vercel/next.js#45715] [Help]
I found a weird behavior when navigating between pages. Usually, when switching pages/routes, I want the scroll position to be set to the top of the page after navigating. Next.js' `Link`'s `scroll` attribute takes care of that.

I'm now in a situation where Next.js is NOT scrolling to the top when switching pages, but I'm not sure whether this should be considered a bug or not.

Here's a reproduction: https://codesandbox.io/s/next-link-scroll-top-bug-f7plkp

I want to have a fixed navbar at the top (see gray area). All the content below (blue area) should be vertically scrollable. I don't want the scollbar to be "on top" of the navbar. So I created a layout where the navbar is fixed and the content area is allowed to grow, which now looks like a typical page:

![image](https://user-images.githubusercontent.com/29319414/217627214-80729ebd-1f24-4e8e-8b44-b3615fcdf69d.png)

The problems is the following: When you scrolled down on one page and after that navigate to another page, the scroll position remains. It does not scroll to the top:

![scrollbug](https://user-images.githubusercontent.com/29319414/217629685-b17a700c-92e5-4cd7-bd15-39ebcf06b78c.gif)

Please note that the text marker is lost on page switches though. 

I am not sure if this is a Next.js bug or not. I can only imagine that this has something to do with React not needing to remount the specific container which wraps the `<Component />` in `_app.tsx`. 

### Hack

My cross-check is the following: I added yet another route, but for that page I force the whole content container (including `<Component />`) to remount. You can see that the scroll resets - I can imagine that's because the `div` is completely fresh in React's world.

```tsx
<div key={asPath === "/force-remount" ? Math.random() + "" : "static"}>
  <Component />
</div>;
```

![scrollbug-force](https://user-images.githubusercontent.com/29319414/217630281-bc1b759e-120a-415b-81fa-5e9fa9072036.gif)

Does someone know what's going on?

---

Btw: Maybe there even is better solution for my layout? I'm glad to hear about other ways so the scrollbar does not appear on the navbar, but the content below is scrollable.

**Comment by gustavostz** [comment]
By default, Next.js does not reset the scroll position when navigating to a new page. To reset the scroll position to the top, you need to implement it yourself. However, forcing a remount every time is not ideal. A better solution would be to use useEffect and useRef to reset the scroll position when navigating to a new page.

Here is an example of this solution, implemented in the _app.tsx file:

```javascript
import React, { useEffect, useRef } from 'react';
import { useRouter } from 'next/router';

function MyApp({ Component, pageProps }) {
  const router = useRouter();
  const mainContentRef = useRef(null);

  useEffect(() => {
    if (mainContentRef.current) {
      mainContentRef.current.scrollTop = 0;
    }
  }, [router.asPath]);

  return (
    <main ref={mainContentRef}>
      <Component {...pageProps} />
    </main>
  );
}

export default MyApp;
```

In this example, the useEffect hook listens for changes to the router.asPath value and sets the scrollTop property of the main content to 0 whenever the path changes. This resets the scroll position to the top of the page whenever the user navigates to a new page.

I also took the liberty of creating a fork and adding this change to the code that you provided: https://codesandbox.io/s/next-link-scroll-top-bug-forked-qc9rpn?file=/pages/_app.tsx

I only made changes to the " _app.tsx" file.

**Reply by bennettdams** [reply]
Thanks, this looks good, gonna try it out.

> By default, Next.js does not reset the scroll position when navigating to a new page

I don't think that is true (or I misunderstand what you mean). This is exactly what the `Link`'s  `scroll` attribute is for. There's even an example in the docs on how to disable that behavior: https://nextjs.org/docs/api-reference/next/link#disable-scrolling-to-the-top-of-the-page

So another question of this post was: Am I doing something weird or is this a bug in Next.js?

**Reply by bennettdams** [reply]
Update: `router.asPath` will also trigger the effect when the the URL search params change, so I use the `router`'s `events` instead:

```tsx
  useEffect(() => {
    router.events.on('routeChangeComplete', () => {
      if (mainContentRef.current) {
        mainContentRef.current.scrollTop = 0
      }
    })
  }, [router.events])
```

**Reply by gustavostz** [reply]
Oh, yes, you are right, I checked the code again and have now figured out the problem. The problem is that the scroll is inside a div (this is the scroll that Next Js does not scroll to the top by default) and not on the page. If you remove the overflowY and set the height in the component, it will work properly, like this:

I updated the fork: https://codesandbox.io/s/next-link-scroll-top-bug-forked-5cnzq3?file=/pages/_app.tsx:1045-1054

I removed the logic I mentioned in my last answer and commented the extra style properties code.

**Reply by bennettdams** [reply]
As mentioned in my original post, the whole point of the layout is to not let the scrollbar overlap the navbar. I find it way more aesthetically pleasing that way. That's why the container containing the page content is y-scrollable.
I know this is a weird setup, just not sure whether this should be considered a bug or not?

**Reply by antoineol** [reply]
Same issue here, for multiple reasons (better control for other libs depending on the scroll, iOS bug for PWA...), I need to set the scroll to an element which is not html. I.e. set html, body, #__next height to 100% and overflow-y: auto on #__next.

Unfortunately, it breaks the scroll reset to top that works otherwise. Unless Next.js has a way to configure the main scrollable element (which is not window.document.scrollingElement anymore).

I'm also interested in a solution or workaround. The router events, why not, but is there a way to know if the "scrollable" parameter was true or false for that event? (We can have both cases in our apps.)

**Reply by ivan-kleshnin** [reply]
> I know this is a weird setup, just not sure whether this should be considered a bug or not?

There's nothing wrong with custom scrollable containers. They are as classic and old as CSS.
NextJS should support scroll restoration besides `body`, for sure. It would probably require some extra app code, not a big deal.
I've posted the same issue to a 3rd party library https://github.com/RevoTale/next-scroll-restorer/issues 
Please upvote if you care.

**Reply by javascripter** [reply]
https://github.com/javascripter/next-scroll-restoration
I've published a new library to support custom scrolling elements in Next.js App Router.
The library does not require native Next.js Scroll Restoration to be disabled either (window scrolling will still be handled by the framework as expected).
Just add `data-scroll-restoration-id` to the element(s) you want to restore scroll positions.

It's still new so I'd appreciate feedback if you try it!
